shared/sbl.js

```diff
@@ -0,0 +1,107 @@
+'use strict'
+
+type TypeFilter = (domain: string, selector: string, data: Array<*>) => ?boolean
+type TypeDomain = {
+  selectors: {[string]: Function},
+  domainFilters: Array<TypeFilter>,
+  selectorFilters: {[string]: TypeFilter},
+  data: Object
+}
+
+var domains: {[string]: TypeDomain} = {}
+var globalFilters: Array<TypeFilter> = []
+var pendingCalls = {} // allows sbl to be called regardless of module load order
+
+function sbl (path: string, ...data?: Array<*>) {
+  var [, name, selector] = /^([^/]+)(.+)/.exec(path)
+  var domain = domains[name]
+  if (!domain) {
+    if (!pendingCalls[name]) pendingCalls[name] = []
+    pendingCalls[name].push([path, data])
+    return
+  }
+  for (let gf of globalFilters) {
+    if (gf(name, selector, data) === false) return
+  }
+  for (let df of domain.domainFilters) {
+    if (df(name, selector, data) === false) return
+  }
+  var sf = domain.selectorFilters[selector]
+  if (sf && sf(name, selector, data) === false) return
+  return domain.selectors[selector].call(domain, ...data)
+}
+
+sbl.registerDomain = function (domain: string, selectors: {[string]: Function}) {
+  if (domains[domain]) throw new Error(`${domain} exists`)
```

Comment:
The code ensures that selectors are unique because they cannot be added afterwards. They can only be specified at the registration step, and the registration step ensures both unique domains and unique selectors.

There is no "selector array", there is only a hashtable. This is necessary for performance reasons.